Duplicati/Library/Main/Operation/DeleteHandler.cs

```diff
@@ -173,45 +174,92 @@ where toDelete.Contains(n.Item2)
         /// </summary>
         /// <returns>The filesets to delete</returns>
         /// <param name="allBackups">The list of backups that can be deleted</param>
-        private DateTime[] GetFilesetsToDelete(DateTime[] allBackups)
+        private DateTime[] GetFilesetsToDelete(Database.LocalDeleteDatabase db, DateTime[] allBackups)
         {
             if (allBackups.Length == 0)
+            {
                 return allBackups;
+            }
+
+            Logging.Log.WriteDryrunMessage(LOGTAG, "***** FileSets", $"List of backup timestamps: {string.Join(", ", allBackups.Select(x => x.ToString()))}");
+
 
             if (allBackups.Select(x => x.ToUniversalTime()).Distinct().Count() != allBackups.Length)
-                throw new Exception(string.Format("List of backup timestamps contains duplicates: {0}", string.Join(", ", allBackups.Select(x => x.ToString()))));
+            {
+                throw new Exception($"List of backup timestamps contains duplicates: {string.Join(", ", allBackups.Select(x => x.ToString()))}");
+            }
 
             List<DateTime> toDelete = new List<DateTime>();
 
-            // Remove backups explicitely specified via option
+            // Remove backups explicitly specified via option
             var versions = m_options.Version;
             if (versions != null && versions.Length > 0)
+            {
                 foreach (var ix in versions.Distinct())
+                {
                     if (ix >= 0 && ix < allBackups.Length)
+                    {
                         toDelete.Add(allBackups[ix]);
+                    }
+                }
+            }
 
             // Remove backups that are older than date specified via option
             var keepTime = m_options.KeepTime;
             if (keepTime.Ticks > 0)
+            {
                 toDelete.AddRange(allBackups.SkipWhile(x => x >= keepTime));
+            }
 
-            // Remove backups via rentention policy option
-            toDelete.AddRange(ApplyRetentionPolicy(allBackups));
+            // Remove backups via retention policy option
+            toDelete.AddRange(ApplyRetentionPolicy(db, allBackups));
 
-            // Check how many backups will be remaining after the previous steps
+            // Check how many full backups will be remaining after the previous steps
             // and remove oldest backups while there are still more backups than should be kept as specified via option
             var backupsRemaining = allBackups.Except(toDelete).ToList();

```

Comment:
I think that's a good idea.  It's not clear in the `FilesetTimes` getter that some other code is depending on the ordering so an uninformed change to a SQL query could lead to disastrous deletion of backup sets.

Reply:
Definitely. And you've made a -really- good find. Unintendended consequences abound from such assumptions in a method. Well done.

We're not programming pace makers but to anyone that loses data, well, the attention to detail is desired.

I really appreciate your attention to such detail which directly impacts the robustness of the code. In that it can then effectively correct future programming mistakes such as your example.

Reply:
Change is in.  It now enforces the sorting.

Reply:
:+1: